### ISSUE TITLE
How to run test script? 

### ISSUE THREAD

**Issue by yimjinkyu1** [epfml/powersgd#8]
import train
train.config["num_epochs"] = 200
train.config["n_workers"] = 8
train.config["rank"] = 0
train.main()

I run above script file.
But It didn't work.

If I change n_workers to 1, then it work.

I think train.py code needs torch.multiprocessing(mp.spawn) or mpirun run script. 
Is it right ?

Please. Could you teach me how to run test script?

**Comment by tvogels** [comment]
Hi @yimjinkyu1,

Indeed, you somehow need to create multiple processes that run code like in your example.

If you use `mpirun`, it could look something like this:

```python
# test.py
import os
import train
train.config["num_epochs"] = 200
train.config["n_workers"] = os.gettenv("OMPI_COMM_WORLD_SIZE")
train.config["rank"] = os.gettenv("OMPI_COMM_WORLD_RANK")
train.config["distributed_init_file"] = "some unique file on a file system workers can all access"

# other configuration like setting output_dir, and logging functions
# ...

train.main()
```

And then run with

`mpirun -n 4 python test.py`

If you use the `slurm` scheduler, you may need to look at other environment variables for `n_workers` and `rank`.
You can also use torch's spawn function and create multiple processes like that, but that wouldn't easily work across different computers.

To get started, you can even make `rank` a command line argument of this script, and start multiple instances manually.

**Comment by bokunwang1** [comment]
According to the following two lines of code, it seems that device "cuda:0" is used in all processes?
https://github.com/epfml/powersgd/blob/26f271a8671748542436685057c19cc0e0fde657/train.py#L74
https://github.com/epfml/powersgd/blob/26f271a8671748542436685057c19cc0e0fde657/mean_accumulator.py#L32

The specs in appendix B of your PowerSGD paper say that there are two GPUs per node. How can it utilize "cuda:1" without specifying the local rank?

**Comment by tvogels** [comment]
I personally use an environment variable

CUDA_VISIBLE_DEVICES=x

To select another GPU x for a specific process.

**Comment by tvogels** [comment]
I'll close this for now. Feel free to re-open.
